File: first_package/script/minimal_node.py
#!/usr/bin/env python3

#!/usr/bin/env python3

import rclpy
from rclpy.node import Node


def main(args=None):
    """
    Main function to initialize a ROS2 node and spin it.

    This function initializes the ROS2 client library, creates a node, logs a greeting message,
    and keeps the node alive until it's manually terminated or receives a shutdown signal.

    Args:
        args (list, optional): A list of command-line arguments passed to the ROS2 node. Defaults to None.
    """
    # Initialize the ROS2 Python client library
    rclpy.init(args=args)

    # Create a new node with the name 'echo_node_py'
    node = Node("echo_node_py")

    # Log a greeting message using the node's logger
    node.get_logger().info(f"Hello from {node.get_name()}")

    # Keep the node alive (listening for callbacks) until it's shut down
    rclpy.spin(node)

    # Shutdown the ROS2 Python client library to clean up resources
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: first_package/src/minimal_node.cpp
/**
 * @file echo_node.cpp
 * @brief Demonstrates a simple ROS 2 node in C++ that logs a message.
 *
 * This program initializes a ROS 2 node, logs a greeting message, and then
 * spins the node to keep it alive until it is manually terminated or receives
 * a shutdown signal.
 */

#include <rclcpp/rclcpp.hpp>  // Include the necessary header to use the ROS 2 C++ client library

/**
 * @brief Main function that initializes a ROS 2 node and logs a message.
 *
 * @param argc Number of command-line arguments.
 * @param argv Array of command-line arguments.
 * @return int Execution status code. Returns 0 on successful execution.
 */
int main(int argc, char* argv[]) {
  // Initialize the ROS 2 system
  rclcpp::init(argc, argv);

  // Create a shared pointer to a new node named "echo_node_cpp"
  auto node = rclcpp::Node::make_shared("echo_node_cpp");

  // Log a message including the node name using the node's logger
  RCLCPP_INFO(node->get_logger(), "Hello from %s", node->get_name());

  // Spin the node so it can perform its work (e.g., processing callbacks)
  rclcpp::spin(node);

  // Shutdown the ROS 2 system before exiting the program
  rclcpp::shutdown();

  return 0;  // Indicate successful execution
}
